lib/patchelf/alt_saver.rb

```diff
@@ -0,0 +1,805 @@
+    attr_reader :in_file # @return [String] Input filename.
+    attr_reader :out_file # @return [String] Output filename.
+
+    # Instantiate a {Saver} object.
+    # @param [String] in_file
+    # @param [String] out_file
+    # @param [{Symbol => String, Array}] set
+    def initialize(in_file, out_file, set)
+      @in_file = in_file
+      @out_file = out_file
+      @set = set
+
+      f = File.open(in_file, 'rb+')
+      @elf = ELFTools::ELFFile.new(f)
+      @buffer = StringIO.new(f.tap(&:rewind).read) # StringIO makes easier to work with Bindata
+
+      @ehdr = @elf.header
+      @segments = @elf.segments # usage similar to phdrs
+      @sections = @elf.sections # usage similar to shdrs
+      update_section_idx!
+
+      # {String => String}
+      # section name to its data mapping
+      @replaced_sections = {}
+      @section_alignment = ehdr.e_phoff.num_bytes
+    end
+
+    # @return [void]
+    def save!
+      @set.each { |mtd, val| send(:"modify_#{mtd}") if val }
+      rewrite_sections
+
+      FileUtils.cp(in_file, out_file) if out_file != in_file
+      patch_out
+      # Let output file have the same permission as input.
+      FileUtils.chmod(File.stat(in_file).mode, out_file)
+    end
+
+    private
+
+    attr_reader :ehdr
+
+    def buf_cstr(off)
+      cstr = []
+      with_buf_at(off) do |buf|
+        loop do
+          c = buf.read 1
+          break if c.nil? || c == "\x00"
+
+          cstr.push c
+        end
+      end
+      cstr.join
+    end
+
+    def buf_move!(dst_idx, src_idx, n_bytes)
+      with_buf_at(src_idx) do |buf|
+        to_write = buf.read(n_bytes)
+        buf.seek dst_idx
+        buf.write to_write
+      end
+    end
+
+    def dynstr
+      find_section '.dynstr'
+    end
+
+    # yields dynamic tag, and offset in buffer
+    def each_dynamic_tags
+      return unless block_given?
+
+      sec = find_section '.dynamic'
+      return unless sec
+
+      shdr = sec.header
+      with_buf_at(shdr.sh_offset) do |buf|
+        dyn = ELFTools::Structs::ELF_Dyn.new(elf_class: shdr.elf_class, endian: shdr.class.self_endian)
+        loop do
+          buf_dyn_offset = buf.tell
+          dyn.clear
+          dyn.read(buf)
+          break if dyn.d_tag == ELFTools::Constants::DT_NULL
+
+          yield dyn, buf_dyn_offset
+          # safety :*
+          buf.seek buf_dyn_offset + dyn.num_bytes
+        end
+      end
+    end
+
+    def find_section(sec_name)
+      idx = find_section_idx sec_name
+      return unless idx
+
+      @sections[idx]
+    end
+
+    def find_section_idx(sec_name)
+      # @sections.find_index { |s| s.name == sec_name }
+      @section_idx_by_name[sec_name]
+    end
+
+    def grow_file(newsz)
+      bufsz = @buffer.size
+      return if newsz <= bufsz
+
+      @buffer.truncate newsz
+    end
+
+    def modify_interpreter
+      @replaced_sections['.interp'] = @set[:interpreter] + "\x00"
+    end
+
+    def modify_needed
+      raise NotImplementedError
```

Comment:
Add a TODO here, to have us (when I revisit here someday in the future) know this is not implemented because of time but not hardness or other reasons.